### PR TITLE
Fixed 9519 -- Added QuerySet.bulk_delete() method that issues minimum queries

### PR DESCRIPTION
**Pull request by Akash-Kumar-Sen**
[Ticket : 9519](https://code.djangoproject.com/ticket/9519)

### CODE REVIEW

File: django/db/models/deletion.py [hunk 1]
```diff
@@ -107,6 +107,8 @@ def __init__(self, using, origin=None):
         # fast_deletes is a list of queryset-likes that can be deleted without
         # fetching the objects into memory.
         self.fast_deletes = []
+        # set perform_signals to false in order to ignore signals
```
Comment (LilyFoote) [inline django/db/models/deletion.py:110]: I think this is sufficiently clear from the name `perform_signals` that we don't need this comment.

File: django/db/models/deletion.py [hunk 2]
```diff
@@ -371,7 +374,10 @@ def collect(
                 # It's something like generic foreign key.
                 sub_objs = field.bulk_related_objects(new_objs, self.using)
                 self.collect(
-                    sub_objs, source=model, nullable=True, fail_on_restricted=False
+                    sub_objs,
+                    source=model,
+                    nullable=True,
+                    fail_on_restricted=False,
```
Comment (LilyFoote) [inline django/db/models/deletion.py:380]: I don't see a reason for this formatting change.

File: django/db/models/query.py [hunk 3]
```diff
@@ -1184,6 +1184,48 @@ def _raw_delete(self, using):
...
+        collector = Collector(using=del_query.db, origin=self, perform_signals=False)
+        collector.collect(del_query)
+        deleted, _rows_count = collector.delete()
+
+        # Clear the result cache, in case this QuerySet gets reused.
+        self._result_cache = None
+        return deleted, _rows_count
+
+    bulk_delete.alters_data = True
+    bulk_delete.queryset_only = True
```
Comment (LilyFoote) [inline django/db/models/query.py]: Given the implementation of this method is identical to that of `delete`, except for setting `perform_signals=False` in `Collector`, I think we should either merge the two methods or delegate the implementation of each to a new private method (`_delete` perhaps).

File: tests/test_bulk_deletion/tests.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    def pre_delete_listener(self, *args, **kwargs):
+        self.pre_delete_signals_count += 1
+
+    def setUp(self):
+        self.post_delete_signals_count = 0
+        self.pre_delete_signals_count = 0
+        pre_delete.connect(self.pre_delete_listener, sender=Baz)
+        post_delete.connect(self.post_delete_listener, sender=Baz)
+
+    def setDown(self):
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:20]:
```suggestion
    def tearDown(self):
```

File: tests/test_bulk_deletion/tests.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        self.post_delete_signals_count = 0
+        self.pre_delete_signals_count = 0
+
+    def test_query_in_bulk_deletion(self):
+        self.pre_delete_signal_was_called = False
+        self.post_delete_signal_was_called = False
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:26]: These are never used.

File: tests/test_bulk_deletion/tests.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            # 2 SELECTS for models with cascaded foreignkey childs
+            # 1 UPDATE for the models with SET NULL foreignkey
+            # 3 DELETES for 3 deleted models
+            foo_objs = Foo.objects.filter(pk=foo.pk)
+            foo_objs.bulk_delete()
+
+        self.assertIsNone(AnotherBar.objects.first().foo)
+        self.assertEqual(self.pre_delete_signals_count, 0)
+        self.assertEqual(self.post_delete_signals_count, 0)
+        self.setDown()
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:45]:
```suggestion
```

File: tests/test_bulk_deletion/tests.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            # 1 SELECT for the model with signals
+            # 1 UPDATE for the models with SET NULL foreignkey
+            # 3 DELETES for 3 deleted models
+            foo_objs = Foo.objects.filter(pk=foo.pk)
+            foo_objs.delete()
+
+        self.assertIsNone(AnotherBar.objects.first().foo)
+        self.assertEqual(self.pre_delete_signals_count, 9)
+        self.assertEqual(self.post_delete_signals_count, 9)
+        self.setDown()
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:69]:
```suggestion
```

File: tests/test_bulk_deletion/tests.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            AnotherBar.objects.create(foo=foo)
+            for j in range(3):
+                Baz.objects.create(bar=bar)
+
+        with self.assertNumQueries(6):
+            # 2 SELECTS for models with cascaded foreignkey childs
+            # 1 UPDATE for the models with SET NULL foreignkey
+            # 3 DELETES for 3 deleted models
+            foo_objs = Foo.objects.filter(pk=foo.pk)
+            foo_objs.bulk_delete()
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:40]:
```suggestion
            Foo.objects.filter(pk=foo.pk).bulk_delete()
```

File: tests/test_bulk_deletion/tests.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            for j in range(3):
+                Baz.objects.create(bar=bar)
+
+        with self.assertNumQueries(7):
+            # 2 SELECTS for models with cascaded foreignkey childs
+            # 1 SELECT for the model with signals
+            # 1 UPDATE for the models with SET NULL foreignkey
+            # 3 DELETES for 3 deleted models
+            foo_objs = Foo.objects.filter(pk=foo.pk)
+            foo_objs.delete()
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:64]:
```suggestion
            Foo.objects.filter(pk=foo.pk).delete()
```

File: tests/test_bulk_deletion/tests.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    def test_query_in_bulk_deletion(self):
+        self.pre_delete_signal_was_called = False
+        self.post_delete_signal_was_called = False
+
+        foo = Foo.objects.create()
+        for i in range(3):
+            bar = Bar.objects.create(foo=foo)
+            AnotherBar.objects.create(foo=foo)
+            for j in range(3):
+                Baz.objects.create(bar=bar)
```
Comment (LilyFoote) [inline tests/test_bulk_deletion/tests.py:33]: It may make sense to move this into `setUp`.

**Review by LilyFoote — CHANGES_REQUESTED** [review]
Looks like a good start to me!

It should be decoupled from the changes for #16907 with a rebase to remove those commits.

**Comment by Akash-Kumar-Sen** [comment]
Thank you for the valuable suggestions @LilyFoote 

> It should be decoupled from the changes for #16907 with a rebase to remove those commits.

This seems redundant as I have already resolved the conflicts. Github will squash all the commits and treat the whole PR as a single commit, Other propositions seems helpful, will update the PR accordingly.

**Comment by felixxm** [comment]
Closing per ticket.

**Pull request closed without merging**